src/services/hsm.service.js

```diff
@@ -0,0 +1,203 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+const { hsm } = require('@dinamonetworks/hsm-dinamo');
+
+const { getConfig } = require('../settings');
+const { lock, lockTypes } = require('../lock');
+const { hsmBusyErrorMessage } = require('../helpers/constants');
+const { HsmError } = require('../errors');
+
+/**
+ * Connects to the HSM and returns the connection object
+ * @returns {Promise<Hsm>}
+ */
+async function hsmConnect() {
+  // Ensures that only one connection will be open at any given moment
+  const canStart = lock.lock(lockTypes.HSM);
+  if (!canStart) {
+    throw new HsmError(hsmBusyErrorMessage);
+  }
+
+  // Gets the connection data from the global config file
+  const config = getConfig();
+  const hsmConnectionOptions = {
+    host: config.hsmHost,
+    authUsernamePassword: {
+      username: config.hsmUsername,
+      password: config.hsmPassword,
+    },
+  };
+
+  // Connects to HSM and returns the connection object
+  return hsm.connect(hsmConnectionOptions);
+}
+
+/**
+ * Disconnects the headless wallet from the HSM.
+ * The connection object is a singleton, so it must not be informed as parameter.
+ * @returns {Promise<void>}
+ */
+async function hsmDisconnect() {
+  await hsm.disconnect();
```

Comment:
Can it be disconnected somehow in the middle of an execution? How do you know that you lost the session connection?

Reply:
Yes, even though it's unlikely, an error can occur in the middle of an execution. In that case we must have good error handling because the exception must be raised and handled by the controller.

This `disconnect()` method is a no-op in case the connection has already been closed or dropped, so it's safe to call on the controller's exception handling.

Reply:
Actually, one of the starting conditions didn't have a protection for this. It was added on [6756809](https://github.com/HathorNetwork/hathor-wallet-headless/pull/356/commits/675680910105fbd54ad684a44c88016d9db1b20c).